src/main.rs

```diff
@@ -0,0 +1,264 @@
+}
+
+type CoinbasePuzzleInst = CoinbasePuzzle<Testnet3>;
+
+fn sample_inputs(
+    degree: u32,
+    rng: &mut (impl CryptoRng + RngCore),
+) -> (EpochChallenge<Testnet3>, Address<Testnet3>, u64) {
+    let epoch_challenge = sample_epoch_challenge(degree, rng);
+    let (address, nonce) = sample_address_and_nonce(rng);
+    (epoch_challenge, address, nonce)
+}
+
+fn sample_epoch_challenge(degree: u32, rng: &mut (impl CryptoRng + RngCore)) -> EpochChallenge<Testnet3> {
+    EpochChallenge::new(rng.next_u32(), Default::default(), degree).unwrap()
+}
+
+fn sample_address_and_nonce(rng: &mut (impl CryptoRng + RngCore)) -> (Address<Testnet3>, u64) {
+    let private_key = PrivateKey::new(rng).unwrap();
+    let address = Address::try_from(private_key).unwrap();
+    let nonce = rng.next_u64();
+    (address, nonce)
+}
+
+#[tokio::main]
+async fn main() -> anyhow::Result<()> {
+    init_logging();
+    let args = Args::parse();
+    info!("Started: {:#}", args);
+
+    let c_info = client_info::ClientInfo::new(&args)?;
+    info!("{:#?}", c_info);
+
+    let rng = &mut thread_rng();
+
+    let max_degree = 1 << 15;
+    let max_config = PuzzleConfig { degree: max_degree };
+    info!("Loading CoinbasePuzzle...");
+    let universal_srs = CoinbasePuzzle::<Testnet3>::setup(max_config).unwrap();
+
+    let degree = (1 << 13) - 1;
+    let config = PuzzleConfig { degree };
+    let coinbase_puzzle = CoinbasePuzzleInst::trim(&universal_srs, config).unwrap();
+    let (epoch_challenge, address, _) = sample_inputs(degree, rng);
+
+    let challenge = Challenge { task_id: "zk_games".to_string(), epoch_challenge: epoch_challenge.clone() };
+
+    let results_file_path = format!("{}_{}", epoch_challenge.clone().epoch_number(), consts::RESULTS_FILE_TEMPLATE);
+    let mut results_file = File::create(results_file_path).unwrap();
+    results_file.write(format!("{:#?}", c_info).as_bytes());
+    results_file.write_all(b"===coinbase_target===");
+    results_file.write_all(&consts::SHARE_TARGET.to_bytes_le().unwrap());
+    results_file.write_all(b"===proof_target===");
+    results_file.write_all(&[0u8]); //
+    results_file.write_all(b"===epoch_challenge===");
+    results_file.write_all(&epoch_challenge.to_bytes_le().unwrap());
+    results_file.write_all(b"===max_degree===");
+    results_file.write_all(&max_degree.to_bytes_le().unwrap());
+    results_file.write_all(b"===degree===");
+    results_file.write_all(&degree.to_bytes_le().unwrap());
+    results_file.write_all(b"===solutions===").unwrap();
+
+    tokio::spawn(do_puzzle(args.clone(), coinbase_puzzle, challenge, address));
+    Ok(())
+}
+
+async fn do_puzzle<N: Network>(
+    args: Args,
+    coinbase_puzzle: CoinbasePuzzle<N>,
+    challenge: Challenge<N>,
+    address: Address<N>,
+) {
+    let threads_num = args.threads_num;
+    let parallel_num = args.parallel_num;
+
+    let mut thread_pools = Vec::new();
```

Comment:
with_capacity is way faster if u know the thread_pools size

Reply:
if the size of array is really big) I think if size isn't really big, it's some nanos

Reply:
oh ok ok